### PR TITLE
Correct, test, and document `FlattenModelDirectives`

### LINKED ISSUE #20425: FlattenModelDirectives does not do scoping
This is a longstanding problem that has only just become urgent for me; I'm creating this issue to associate with my work to fix it.

`FlattenModelDirectives` has an implementation comment `// NOTE: Does not handle scoping!`.  This is accurate:  Flattened model directives lack the scoping requested by the original directives, and so are typically invalid for any model with cross-references or duplicate models.

It also has no documentation comment, nor is it tested.

In fact, to the extent that `FlattenModelDirectives` is useful (to make a durable, self-contained record of a scenario-as-executed; to create non-colliding content for hashing), it must be the case that processing model directives in deep form and flattened form should result in the same model instances.  As written it does not achieve that.

Fixing this probably also entails making the assignment of scoped names more explicit.

### PR DESCRIPTION
**Pull request by ggould-tri**
FlattenModelDirectives did not handle scoped names, which made it difficult to use in scenarios with multiple instances of the same model.

It was also undocumented and untested.

This change adds correct namespace handling; it also adds documentation and a minimal test (minimal because any naming-related errors are caught during the assembly of the MultibodyTree, and because the scoped test fixtures were designed to exercise naming).

 * Fixes #20425.
 * Removes misleading comments (including a mention of RigidBodyTree!).
 * Documents FlattenModelDirectives.
 * Documents the special-case naming semantics in add_frame directives.

- - -
This change is [Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)

### CODE REVIEW

**Review by ggould-tri — COMMENTED** [review]
+@IanTheEngineer as the initial reporter.  This one is a delightful educational voyage full of discovery and wonder!

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-Nhhfwpm9N5DI-vni4KZ:b-tbc2gg)* status: LGTM missing from assignee IanTheEngineer, needs platform reviewer assigned, needs at least two assigned reviewers, missing label for release notes (waiting on @ggould-tri)

**Review by ggould-tri — COMMENTED** [review]
+(release notes: fix) 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-NhhkkVq8ILNJwfBct-w:b-chsgxs)* status: LGTM missing from assignee IanTheEngineer, needs platform reviewer assigned, needs at least two assigned reviewers

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 12 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5I1Cy1jAwCEA2hmT4):*
Everything after ", which ..." here is overly broad and therefore wrong.  There are _plenty_ of composed models that neither use nor require namespacing.
___
*[`multibody/parsing/process_model_directives.h` line 32 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5IQnV9eiScNpEO5y9:-Ni5IQnV9eiScNpEO5yA:bb7phvt) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.h#L32)):*
> ```objc
> /// Flatten model directives into a single object.
> ///
> /// This method removes all AddDirectives directives from the given
> ```

typo



_Suggestion:_
```objc
function
```
___
*[`multibody/parsing/process_model_directives.cc` line 40 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5MOr2-Ca8YyuaZI4X:-Ni5MOr2-Ca8YyuaZI4Y:b7cuum9) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L40)):*
> ```cpp
> }
> 
> std::string NamespaceJoin(std::string prefix, std::string name) {
> ```

nit Passing string-by-value without std::move is an (unconventional) pessimization.  It should either be a string_view by-value, or a string by-const-ref.
___
*[`multibody/parsing/process_model_directives.cc` line 44 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5M9DW5nLWFiuEFEJp:-Ni5M9DW5nLWFiuEFEJq:b7cuum9) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L44)):*
> ```cpp
> }
> 
> AddWeld ApplyDirectiveNamespace(const AddWeld& orig, std::string prefix) {
> ```

nit Passing string-by-value without std::move is an (unconventional) pessimization.  It should either be a string_view by-value, or a string by-const-ref.
___
*[`multibody/parsing/process_model_directives.cc` line 48 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5QD8D5pv82PwqGyM0:-Ni5QD8D5pv82PwqGyM1:b-b1w6qa) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L48)):*
> ```cpp
>   AddWeld result = orig;
>   result.parent = NamespaceJoin(prefix, orig.parent);
>   result.child = NamespaceJoin(prefix, orig.child);
> ```

This doesn't seem right.  Consider if the `parent` or `child` were `"world"`, we don't want `left::world` here.

Best guess, we need to match the logic of `ParseModelDirectivesImpl` helper lambda `get_scoped_frame`, which might be as simple as special-casing `world` here.
___
*[`multibody/parsing/process_model_directives.cc` line 73 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5P9voDG3t0RfIXCzg:-Ni5P9voDG3t0RfIXCzh:b9clt68) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L73)):*
> ```cpp
>   AddFrame result = orig;
>   result.name = NamespaceJoin(prefix, orig.name);
>   result.X_PF.base_frame = NamespaceJoin(prefix, *orig.X_PF.base_frame);
> ```

This doesn't seem right.  Consider if the base_frame were `world`, we don't want `left::world` here.

Best guess, we need to match the logic of `ParseModelDirectivesImpl` helper lambda `get_scoped_frame`, which might be as simple as special-casing `world` here.
___
*[`multibody/parsing/process_model_directives.cc` line 81 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5Qt4e7n6rOi0z8qtu:-Ni5Qt4f5VmK-VWPmiXm:b-pugt42) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L81)):*
> ```cpp
>   if (prefix.empty()) return orig;
>   AddCollisionFilterGroup result = orig;
>   result.name = NamespaceJoin(prefix, orig.name);
> ```

The `AddCollisionFilterGroup` member field `name` is documented as "must not be a scoped name".

I don't immediately see how `ApplyDirectiveNamespace` can do anything other than throw a "not implemented" error when prefix is non-empty.  Maybe there is some way around it, but in any case you'll need to add more unit test coverage of this item if you want to support it.
___
*[`multibody/parsing/process_model_directives.cc` line 98 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5SA1184Vjx8fE8q09:-Ni5SA1184Vjx8fE8q0A:b2bihgy) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L98)):*
> ```cpp
>     const PackageMap& package_map,
>     ModelDirectives* out,
>     const std::string& prefix) {
> ```

Throughout -- we have a variable `prefix` with specific semantics.  I _think_ from reading the other dmd code that `prefix` is really the same as `model_namespace` used elsewhere.  If I'm right that it's the same concept, then we should use the same name for it universally for clarity.
___
*[`multibody/parsing/process_model_directives.cc` line 102 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5RprQ7uIwcTibjOfE:-Ni5RprRE2YPaCp7DJ97:b-h1wqlc) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L102)):*
> ```cpp
>     if (directive.add_directives) {
>       const AddDirectives& sub = *directive.add_directives;
>       const auto sub_file = ResolveModelDirectiveUri(sub.file, package_map);
> ```

nit Too-obscure use of `auto`.  The return type is not apparent from the RHS.
___
*[`multibody/parsing/process_model_directives.cc` line 104 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5LbZC7Aky36Qf-zxU:-Ni5LbZC7Aky36Qf-zxV:bzdxgjf) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L104)):*
> ```cpp
>       const auto sub_file = ResolveModelDirectiveUri(sub.file, package_map);
>       FlattenModelDirectivesInternal(
>         LoadModelDirectives(sub_file), package_map, out,
> ```

nit This indentation is wrong.  Function argument are indented at least +4 from the start of their function name.

(Probably simplest to just hit this file with git clang-format-diff.  There are several more mis-indentations.)
___
*[`multibody/parsing/process_model_directives.cc` line 125 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5SX6gByOY_AFGAf4V:-Ni5SX6gByOY_AFGAf4W:bgrlj6h) ([raw file](https://github.com/RobotLocomotion/drake/blob/3590141447c4675bb659179a57e49b1e4aecaa2d/multibody/parsing/process_model_directives.cc#L125)):*
> ```cpp
>       // AddFrame has a subtle alteration of the scoped name semantics.  In
>       // the event that the frame name has no namespace, it gets the namespace
>       // of the base frame of its X_PF offset.
> ```

nit It smells weird to me to have application logic inside the if-else chain.  I would expect it to be clearer if this if-else chain was all just boilerplate of calling per-struct helper functions, and this "subtle" logic (and its comments) could move into the helper function.

**Review by ggould-tri — COMMENTED** [review]
checkpoint, still working 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-NiBbKrz1MdBbFOG421_:b-37h708)* status: 6 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5I1Cy1jAwCEA2hmT4):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Everything after ", which ..." here is overly broad and therefore wrong.  There are _plenty_ of composed models that neither use nor require namespacing.
</blockquote></details>

Done.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 2 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 5 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

**Review by ggould-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 5 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Nikjdqg1nFPpmyMp_Hb:-Nikjdqg1nFPpmyMp_Hc:b-bvqpp8) (no related file):*
Working:   Needs a keyhole test of the special case of the frame named "world".

___
___
*[`multibody/parsing/process_model_directives.cc` line 48 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5QD8D5pv82PwqGyM0:-NiedCWe5p2baYxI80Jj:bg4diuk) ([raw file](https://github.com/RobotLocomotion/drake/blob/36c07b6c39c7bb963a251b9efe12e1a909c22b68/multibody/parsing/process_model_directives.cc#L48)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This doesn't seem right.  Consider if the `parent` or `child` were `"world"`, we don't want `left::world` here.

Best guess, we need to match the logic of `ParseModelDirectivesImpl` helper lambda `get_scoped_frame`, which might be as simple as special-casing `world` here.
</blockquote></details>

`get_scoped_frame` itself can't be hoisted since it's using a plant as its data store, so I've hoisted the logic to `NamespaceJoin` above (since the specialness of "world" is a fact about our namespace algebra).
___
*[`multibody/parsing/process_model_directives.cc` line 73 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5P9voDG3t0RfIXCzg:-Nieeke92uPKguZPinml:b-q64pkt) ([raw file](https://github.com/RobotLocomotion/drake/blob/36c07b6c39c7bb963a251b9efe12e1a909c22b68/multibody/parsing/process_model_directives.cc#L73)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This doesn't seem right.  Consider if the base_frame were `world`, we don't want `left::world` here.

Best guess, we need to match the logic of `ParseModelDirectivesImpl` helper lambda `get_scoped_frame`, which might be as simple as special-casing `world` here.
</blockquote></details>

See above.
___
*[`multibody/parsing/process_model_directives.cc` line 81 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5Qt4e7n6rOi0z8qtu:-Nief2uX4uN7K0q3k7g-:bdcv6l8) ([raw file](https://github.com/RobotLocomotion/drake/blob/36c07b6c39c7bb963a251b9efe12e1a909c22b68/multibody/parsing/process_model_directives.cc#L81)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The `AddCollisionFilterGroup` member field `name` is documented as "must not be a scoped name".

I don't immediately see how `ApplyDirectiveNamespace` can do anything other than throw a "not implemented" error when prefix is non-empty.  Maybe there is some way around it, but in any case you'll need to add more unit test coverage of this item if you want to support it.
</blockquote></details>

Good point.
___
*[`multibody/parsing/process_model_directives.cc` line 98 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5SA1184Vjx8fE8q09:-NiehRIJD7DvEnWuc-lv:b4n8z2u) ([raw file](https://github.com/RobotLocomotion/drake/blob/36c07b6c39c7bb963a251b9efe12e1a909c22b68/multibody/parsing/process_model_directives.cc#L98)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Throughout -- we have a variable `prefix` with specific semantics.  I _think_ from reading the other dmd code that `prefix` is really the same as `model_namespace` used elsewhere.  If I'm right that it's the same concept, then we should use the same name for it universally for clarity.
</blockquote></details>

I think that you are right, although I hadn't been thinking about it that way:  Every nonleaf in the naming tree is a model instance.
___
*[`multibody/parsing/process_model_directives.cc` line 125 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5SX6gByOY_AFGAf4V:-NikjQ579wUBaFDHntD0:buhydf9) ([raw file](https://github.com/RobotLocomotion/drake/blob/36c07b6c39c7bb963a251b9efe12e1a909c22b68/multibody/parsing/process_model_directives.cc#L125)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit It smells weird to me to have application logic inside the if-else chain.  I would expect it to be clearer if this if-else chain was all just boilerplate of calling per-struct helper functions, and this "subtle" logic (and its comments) could move into the helper function.
</blockquote></details>

This was surprisingly difficult to factor correctly but I believe it is now OK.

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 6 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5I1Cy1jAwCEA2hmT4):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The old commit message has snuck back in here again.

Actually, looking more closely, all of r1->r2 got dropped and needs to be resurrected.

**Comment by ggould-tri** [comment]
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5I1Cy1jAwCEA2hmT4):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The old commit message has snuck back in here again.

Actually, looking more closely, all of r1->r2 got dropped and needs to be resurrected.
</blockquote></details>

dangit.  Working.

**Comment by ggould-tri** [comment]
*[`-- commits` line 6 at r1](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Ni5I1Cy1jAwCEA2hmT4):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

dangit.  Working.
</blockquote></details>

r2->r4 should now be what r2->r3 was meant to be...  unless I messed up the merge again...

**Review by jwnimmer-tri — COMMENTED** [review]
BTW @IanTheEngineer please re-review the new changes.

> Reviewed 1 of 4 files at r2, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-NikybYN-y8xxmx9exa3:b-gcfquw)* status: 7 unresolved discussions, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`multibody/parsing/process_model_directives.cc` line 52 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NiksuuTC1uQB0XV8CLv:-NiksuuTC1uQB0XV8CLw:b-o7qsza) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L52)):*
> ```cpp
> 
> AddWeld ApplyDirectiveNamespace(const AddWeld& orig,
>                                 std::string model_namespace) {
> ```

nit This `model_namespace` should be pass-by-const-ref.
___
*[`multibody/parsing/process_model_directives.cc` line 79 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikwLfo80Fn8FWv3Y2p:-NikwLfo80Fn8FWv3Y2q:b-sam28w) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L79)):*
> ```cpp
>                                  const std::string& model_namespace) {
>   AddFrame result = orig;
>   const ScopedName scoped_name = ScopedName::Parse(orig.name);
> ```

BTW I probably would have been able to grok this function more easily if either this variable were called `orig_scoped_name` to help remind me where it came from, or it was moved down to just before its first point-of-use.
___
*[`multibody/parsing/process_model_directives.cc` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikvdTZBjwA-ndcfgnx:-NikvdTZBjwA-ndcfgny:b37z73d) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L84)):*
> ```cpp
>     result.name = NamespaceJoin(model_namespace, orig.name);
>     result.X_PF.base_frame =
>       NamespaceJoin(model_namespace, *orig.X_PF.base_frame);
> ```

nit Here the `operator*` is a segfault when the user has forgotten to check `directives.IsValid()` before asking for flattening.  At minimum we should probably be using `value()` instead of the operator here, so at least we get some exception.
___
*[`multibody/parsing/process_model_directives.cc` line 94 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Nikwc0n2ZVIPIDA2lef:-Nikwc0n2ZVIPIDA2leg:bq42jwd) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L94)):*
> ```cpp
>     // base frame's namespace prepended.
>     const std::string base_frame_namespace(
>       ScopedName::Parse(*orig.X_PF.base_frame).get_namespace());
> ```

nit Wrong indentation (+4 for expressions).  Ditto two lines later.  (In short, just run clang-format-diff again.)
___
*[`multibody/parsing/process_model_directives.cc` line 109 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikuKm5BlImr0JHa9d1:-NikuKm64i_MddSQuNyx:b-f0tkb1) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L109)):*
> ```cpp
>   // TODO(ggould-tri) The rest of the structure could concievably be salvaged
>   // rather than failing outright, but I don't yet see a sound use case.
>   DRAKE_THROW_UNLESS(model_namespace.empty());
> ```

nit This is fine, but now we probably need to update the commit message with something like: "FlattenModelDirectives will not notice with a namespace has been attempted to be added to a collision filter group and will fail-fast instead of ignoring it".

(In the alternative, this could be a `drake::log()->warn` if you want to split the difference.)
___
*[`multibody/parsing/README_model_directives.md` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikxYad6PBxe6Sn07yb:-NikxYae048806HF8pPe:b-etnjh7) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/README_model_directives.md#L84)):*
> ```Markdown
>   `top_level::my_model`, the frame is `my_frame`.
> 
> All element names will be automatically prefixed with that model
> ```

The r1 text made perfect sense to me, but the r2 text is really confusing, because it has lost its introductory sentence along the lines of "Here's how scoping interacts with  `add_directives`".

**Review by ggould-tri — COMMENTED** [review]
This PR just failed to detect a rebase from master.  I may need to reopen it... pending

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-Niq7y3d34s0iK8_cE2v:bvzvelr)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform), commits need curation (https://drake.mit.edu/reviewable.html#curated-commits) (waiting on @ggould-tri)

___
*[`multibody/parsing/process_model_directives.cc` line 79 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikwLfo80Fn8FWv3Y2p:-Niq-ALuB90ATyFuOQQz:bhm4ad4) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L79)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I probably would have been able to grok this function more easily if either this variable were called `orig_scoped_name` to help remind me where it came from, or it was moved down to just before its first point-of-use.
</blockquote></details>

Or, why not both :-)
___
*[`multibody/parsing/process_model_directives.cc` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikvdTZBjwA-ndcfgnx:-Niq-Cgy25q9mf31rjHT:bhm4ad4) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Here the `operator*` is a segfault when the user has forgotten to check `directives.IsValid()` before asking for flattening.  At minimum we should probably be using `value()` instead of the operator here, so at least we get some exception.
</blockquote></details>

Or, why not both :-)
___
*[`multibody/parsing/process_model_directives.cc` line 94 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Nikwc0n2ZVIPIDA2lef:-Niq1yVNE7KhYFf-Jq2Y:b-8qr1w5) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L94)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

nit Wrong indentation (+4 for expressions).  Ditto two lines later.  (In short, just run clang-format-diff again.)
</blockquote></details>

For whatever reason clang-format was a lot more opinionated this time.  Minor tweaks throughout the file, though mostly not defective.  shrug-emoji.
___
*[`multibody/parsing/README_model_directives.md` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikxYad6PBxe6Sn07yb:-Niq3c1QDunWZ-obe3oZ:bkbqzhe) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/README_model_directives.md#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

The r1 text made perfect sense to me, but the r2 text is really confusing, because it has lost its introductory sentence along the lines of "Here's how scoping interacts with  `add_directives`".
</blockquote></details>

I've tried to reword this to compromise between your and Ian's preferences.  I hope this version is equally hateful :-)

**Review by ggould-tri — COMMENTED** [review]
... seems to have fixed itself.  No more r16 anyway.  :shrug: 

> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-Niq8o7T3iOyco0ItY6d:b-a1eeqx)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform)

**Review by jwnimmer-tri — COMMENTED** [review]
> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`multibody/parsing/process_model_directives.cc` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikvdTZBjwA-ndcfgnx:-Niq9SCrBJb6quAtvfxk:bnyntq0) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L84)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Or, why not both :-)
</blockquote></details>

You fixed the one below (on line 95) but not this one (line 83).
___
*[`multibody/parsing/process_model_directives.cc` line 94 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-Nikwc0n2ZVIPIDA2lef:-Niq92sCBSePB9Uxerni:b4qws9z) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L94)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

For whatever reason clang-format was a lot more opinionated this time.  Minor tweaks throughout the file, though mostly not defective.  shrug-emoji.
</blockquote></details>

Hmm, did you reformat the whole file instead of just the diffs?

(By default, the `multibody/parsing` directory is not yet clang-formatted overall.)

**Review by ggould-tri — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438)* status: 1 unresolved discussion, LGTM missing from assignee jwnimmer-tri(platform)

___
*[`multibody/parsing/process_model_directives.cc` line 84 at r4](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NikvdTZBjwA-ndcfgnx:-NiqDsnOAl2IeJW0n3v0:b-vmasq5) ([raw file](https://github.com/RobotLocomotion/drake/blob/60f8a143d5a5c4a79121a0f3ca378fd11b046654/multibody/parsing/process_model_directives.cc#L84)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

You fixed the one below (on line 95) but not this one (line 83).
</blockquote></details>

Done.  Given the previous revisionstorm, I must have been editing a different version of this file from head, probably fallout from the previous accidental revert.

**Review by jwnimmer-tri — COMMENTED** [review]
:lgtm: platform.

> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-NiqFIQmADQumepxw1ZF:b-vtj4vo)* status: 1 unresolved discussion

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NiqFIy-6smXgrIF8JXS:-NiqFIy-6smXgrIF8JXT:brlu5dv) (no related file):*
Blocking comment to wait for @IanTheEngineer's final stamp.

___

**Comment by svenevs** [comment]
@drake-jenkins-bot mac-arm-ventura-clang-bazel-experimental-release please.

maybe it will work, CI is broken

**Review by IanTheEngineer — COMMENTED** [review]
:lgtm: feature. I did some light local re-testing of your updated branch against some `model_namespace`'d directives and it flattened and namespaced wonderfully. Thanks for iterating, Grant.

> Reviewed 4 of 4 files at r4, 1 of 3 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-:-Niu4Jpi8gNjYCasfNRl:bamlge7)* status: :shipit: complete! all discussions resolved, LGTM from assignees jwnimmer-tri(platform),IanTheEngineer

*[a discussion](https://reviewable.io/reviews/RobotLocomotion/drake/20438#-NiqFIy-6smXgrIF8JXS:-Niu4FtD9PXAZV8ghBGZ:b-c2i3a4) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Blocking comment to wait for @IanTheEngineer's final stamp.
</blockquote></details>

Stamped!

___
